src/array_buffer.rs

```diff
@@ -246,25 +246,50 @@ pub type BackingStoreDeleterCallback = unsafe extern "C" fn(
   deleter_data: *mut c_void,
 );
 
-pub unsafe extern "C" fn boxed_slice_deleter_callback(
-  data: *mut c_void,
-  byte_length: usize,
-  _deleter_data: *mut c_void,
-) {
-  let slice_ptr = ptr::slice_from_raw_parts_mut(data as *mut u8, byte_length);
-  let b = Box::from_raw(slice_ptr);
-  drop(b);
+pub(crate) mod sealed {
+  pub trait Rawable<T: ?Sized> {
+    fn into_raw(self) -> *const ();
+    unsafe fn drop_raw(ptr: *const (), size: usize);
+  }
 }
 
-pub unsafe extern "C" fn vec_deleter_callback(
-  data: *mut c_void,
-  byte_length: usize,
-  deleter_data: *mut c_void,
-) {
-  let capacity = deleter_data as usize;
-  drop(Vec::from_raw_parts(data as *mut u8, byte_length, capacity))
+impl sealed::Rawable<[u8]> for Vec<u8> {
+  unsafe fn drop_raw(ptr: *const (), size: usize) {
+    <Box<[u8]> as sealed::Rawable<[u8]>>::drop_raw(ptr, size);
+  }
+
+  fn into_raw(self) -> *const () {
+    self.into_boxed_slice().into_raw()
+  }
 }
 
+macro_rules! rawable {
+  ($container:ident) => {
+    impl<T: Sized> sealed::Rawable<T> for $container<T> {
+      fn into_raw(self) -> *const () {
+        Self::into_raw(self) as _
+      }
+
+      unsafe fn drop_raw(ptr: *const (), _len: usize) {
+        _ = Self::from_raw(ptr as _);
+      }
+    }
+
+    impl sealed::Rawable<[u8]> for $container<[u8]> {
+      fn into_raw(self) -> *const () {
+        Self::into_raw(self) as _
+      }
+
+      unsafe fn drop_raw(ptr: *const (), len: usize) {
+        _ = Self::from_raw(ptr::slice_from_raw_parts_mut(ptr as _, len));
+      }
+    }
+  };
+}
+
+// Implement Rawable for single-ownership container types
+rawable!(Box);
```

Comment:
Left the macro as this might make sense to implement for `Rc<[u8]>` and `Arc<[u8]>` if they are single-ownership at the time.